enhancements/big-bundle-unpacking.md

```diff
@@ -0,0 +1,138 @@
+
+As an operator author, I want OLM to be capable of unpacking and installing a bundle that contains an arbitrary number of resource manifests, each no larger than the limit for an individual API resource on a target cluster, so that the size of what I can package as an operator is no more limited than that target cluster.
+
+### Implementation Details/Notes/Constraints
+
+#### PodSpec Pull Requirement
+
+In order to support pulling bundle images on OpenShift, OLM must adhere to the platform's [pull policies](https://docs.openshift.com/container-platform/4.5/openshift_images/image-configuration.html). The lowest risk way to do this is to rely on CRI to pull images and apply the policies opaquely, and the most straightforward initiate a pull is via a PodSpec. This means that, when a bundle image is pulled and unpacked by OLM today, a `Job` is created which specifies three containers:
+
+- a bundle util init container
+  - contains a statically linked bundle tool that can search the container’s filesystem for bundle metadata -- annotations.yaml -- and copy the referenced manifest files to a bundle volume
+  - mounts a single bundle util volume
+  - at runtime, uses cp to copy this bundle tool to the bundle util volume
+- a bundle data init container
+  - pull spec is used to delegate pulling of the bundle image via CRI-O
+  - mounts two volumes: bundle util and bundle data
+  - at runtime, executes the bundle tool from the bundle util mount to copy the bundle to the bundle data mount
+- a bundle extract container
+  - contains opm
+  - mounts the bundle data volume
+  - runs `opm extract` against the bundle data mount which results in a `ConfigMap` being created that contains the bundle’s kubectl-able content
+
+#### Partitioning and Storage
+
+Instead of extracting bundle content directly to a `ConfigMap` -- as outlined in the [previous section](#podspec-pull-requirement) -- the unpack `Job` will first encode it as JSON and compress the result. The compressed JSON will then be split into chunks of a [fixed size](#determining-chunk-size), each wrapped with a header to keep track of its position in the chunk sequence. Each chunk is then encoded as JSON and stored in a respective `ConfigMap`. To each of these `ConfigMaps` a `porcelain.operators.coreos.com/bundle: <bundle-name>` label is applied to faciliate later retrieval. A `porcelain.operators.coreos.com/signature: <signature>` annotation is also generated for each chunk and is used to validate its integrity. When the content for a particular bundle is then retrieved, the `ConfigMaps` with the matching bundle name label are collected, sorted by the position tracked in each header, reconstituted into a single artifact, and decompressed before being returned for a `/content` request. If the signature verification fails for any chunk, a status condition indicating will be added to the associated bundle which will trigger a new unpack `Job` to run in a bid to rectify the issue.
```

Comment:
my guess is that the majority of individual manifests (I have no data, just guessing) will fit within a single configmap. could we reserve the chunking for manifests that don't fit?

Then unpacking a large bundle could look like:

ConfigMap1
 - service
 - serviceaccount
 - rbac

ConfigMap2
 - csv
 - CRD A

ConfigMap3
 - CRD B Part 1

ConfigMap4
 - CRD B Part 2

This seems like it will increase the debuggability of bundle unpacking for the common case

Reply:
hmm, I don't disagree that it would be more human readable, but I would like the storage mechanism -- in this case, `ConfigMaps` -- to remain opaque to users. I think it simplifies the implementation to treat data as cattle.

Reply:
That looks good to me - I'm not sure I'm seeing the distinction as writ though.

will the configmap look like

```yaml
data:
   {service}{serviceaccount}{rbac}{CRDA}{CSVpart1}
---
data:
  {CSVPart2}
```

or like:

```yaml
data:
   - {service}
   - {serviceaccount}
   - {rbac}
   - {CRDA}
   - {CSVpart1}
---
data:
  - {CSVPart2}
```

Reply:
Ah, I see now. I had envisioned the former, but the latter may be more clear. I do think this will be opaque to most end users, but I see your point.